### PR TITLE
Fix reused attributes and update to use InSpec 3.x patterns. Fixed user guidance for best security practice.

### PR DESCRIPTION
**Pull request by aaronlippold**
- addresses reuse of centralized inputs(attributes)
- removes use of constants in controls in favor of instance variables
- moves all core inputs (attributes) to the `inspec.yml`
- updated the `README.md` to address these changes
- updated the `README.md` to use `env_vars` for sensitive data (passwords etc)
- updated the `inspec.yml` to use `env_vars` so they are never stored in plain text
- update the `README.md` to not suggest users store password in plain text
- updated the user running example to use the more modern `reporters` ( always at the end of the line )
- updated the user reporting examples to use the `html` reporter, `json` reporter and `cli' reporter.
- update the `README.md` to reference `heimdall-lite` so users can easily review their `json` results.

Closes #21
Overrides #21 

Signed-off-by: Aaron Lippold <lippold@gmail.com>

### CODE REVIEW

File: inspec.yml [hunk 1]
```diff
@@ -9,10 +9,233 @@ license: Apache 2.0
...
+inspec_version: "~> 3.0"
+
+# Note: Export the <%= ENV['pg_owner'] %> to the envoriment to set them
+# at runtime.
+
+attributes:
+  - name: pg_owner
+    description: 'The system user of the postgres process'
+    type: string
+    value: <%= ENV['PG_OWNER'] %>
```
Comment (chris-at-work) [inline inspec.yml:29]: So we need to set ENVs when running Inspec?  `PG_OWNER=someone bundle exec inspec exec ...`
Or a downstream project could use `dotenv` or similar to define environments?  Or maybe you set these as a user's profile on a "scanning box"?  I'm fairly clear on how to do these things, curious what your preference is or what made you switch from string defaults in the YAML.

File: README.md [hunk 2]
```diff
@@ -31,64 +31,117 @@ The PostgreSQL STIG provides guidance on the configuration of PostgreSQL to addr
...
 inspec --version
 ```
 
-### Setting attributes.yml
+### Setting & Reviewing the Core Profile Attributes
+
+The `core` or `shared` attributes are set in the `inspec.yml` file in the profile. This stores the default values for the
+shared attributes that the profile uses to examine your system.
+
+These attributes _should not be edited directly_!
```
Comment (jkatz) [inline README.md:106]: I would make it

> These attributes _**should not be edited directly**_!

as this sounds pretty important 😄

File: README.md [hunk 3]
```diff
@@ -31,64 +31,117 @@ The PostgreSQL STIG provides guidance on the configuration of PostgreSQL to addr
...
+
+If you need to over-ride the default values for the core attributes to better match your system under evaluttaion, please:
```
Comment (jkatz) [inline README.md]: > over-ride
to
> override

> evaluttaion
to
> evaluation

File: README.md [hunk 4]
```diff
@@ -31,64 +31,117 @@ The PostgreSQL STIG provides guidance on the configuration of PostgreSQL to addr
...
+
+1.  Review the attributes and their defaults in the `inspec.yml` and note the ones that you need to enhance.
+2.  Create a `system` or `instance` specific `attributes.yml` - such as `attributes.mysystem.yml` and over-ride the attributes with this file.
```
Comment (jkatz) [inline README.md]: 1. We can just use `1.` on each line as its markdown and it will autonumber
1. `over-ride` => `override`

File: README.md [hunk 5]
```diff
@@ -31,64 +31,117 @@ The PostgreSQL STIG provides guidance on the configuration of PostgreSQL to addr
...
+3.  When you run the profile, you can load your updated attributes using the `--attrs` flag on the `inspec exec` command (see below)
+
+### Sensitive Attributes (aka `passwords`, `system accounts` and `owners` in your Postgres database )
```
Comment (jkatz) [inline README.md]: Our preference has been to use `PostgreSQL` when describing the database, not `Postgres`

File: README.md [hunk 6]
```diff
@@ -31,64 +31,117 @@ The PostgreSQL STIG provides guidance on the configuration of PostgreSQL to addr
...
+
+It is _**NOT RECOMMENDED**_ that you store sentive information - such as passwords - in plain text yaml files, so the profile is able to use `environment vairables` that are set on the \*`inspec runner` - aka the host you are running the `inspec exec` command from.
+
+You _will need to set these environment variables_ or you _could_ set them in your `attributes.mysystem.yml` but we _would prefer_ you set them via your `environment` as it is a better practice.
```
Comment (jkatz) [inline README.md]: I'd suggest rephrasing it:

> The recommend way to store sensitive information is to use one of the environmental variables outlined below. For instance, you can set the password for the PostgreSQL user with the `PG_USER_PWD` environmental variable.
>
> Alternatively, you can set these variables in your `attributes.mystem.yml` file, but again, this is not recommended for sensitive information like passwords.

File: README.md [hunk 7]
```diff
@@ -111,11 +164,11 @@ pg_shared_dirs: [
 pg_conf_mode: '0600'
 pg_ssl: 'on'
 pg_log_dest: 'syslog'
-pg_syslog_facility: ['local0']
+pg_syslog_facility: ['local0
```
Comment (jkatz) [inline README.md]: I think this is missing a `]`

File: README.md [hunk 8]
```diff
@@ -111,11 +164,11 @@ pg_shared_dirs: [
...
 pg_syslog_owner: 'postgres'
 
-pgaudit_log_items: ['ddl','role','read','write']
-pgaudit_log_line_items: ['%m','%u','%c']
+pgaudit_log_items: ['ddl','role','read','write
```
Comment (jkatz) [inline README.md]: This and the line below seems to be missing a `]`

**Review by jkatz — CHANGES_REQUESTED** [review]
Blocking changes involve missing `]` on some examples with YAML arrays. Some suggestions on phrasing.

File: README.md (hunk 2 above)
Comment (aaronlippold) [reply README.md:106]: resolved

File: README.md (hunk 3 above)
Comment (aaronlippold) [reply README.md]: resolved

File: README.md (hunk 4 above)
Comment (aaronlippold) [reply README.md]: resolved

File: README.md (hunk 5 above)
Comment (aaronlippold) [reply README.md]: resolved

File: README.md (hunk 6 above)
Comment (aaronlippold) [reply README.md]: resolved

File: README.md (hunk 7 above)
Comment (aaronlippold) [reply README.md]: resolved

**Comment by jkatz** [comment]
@aaronlippold Read through as much as I could -- looks good to me!

My only question is if the removal of "severity" from each of the control files is deliberate (well, I'm guessing it is).

Other than that, I am ready to approve this PR.

**Review by hunleyd — COMMENTED** [review]
same question as @jkatz about the severity removal, but otherwise lgtm

**Comment by aaronlippold** [comment]
The original intention of removing the severity tag was that we felt it was repeated data as we have the impact with value for the control. Given that we could easily calculate high medium and low from the impact value.  

However with recently come across an issue and inspect tools where the conversion of the inspect results back to a checklist file could actually benefit from the severity tag as it is the original low medium high value of the control regardless of changes to the impact value of the control due to overlays or dependent profiles. 

So I might actually suggest keeping the severity tag. The corner case comes in when I set a control in an overlay to impact zero AKA making it not applicable. With the removal of the severity tag and the fact that the impact was set to zero we actually lose what the original high medium and low value is for those data values. I am going to review the Json data and see if the parent profile still has the proper impact and I can retrieve it from there. In that case I don't lose any fidelity of data and I don't add any unnecessary data in the form of tags. if the data really is lost for whatever reason I will suggest putting the tags back in.

**Comment by sharmay** [comment]
@aaronlippold while testing this PR I am getting this error  `Input 'pg_owner' with value '' does not validate to type 'String'.` 

attributes.yml  has `pg_owner: 'postgres'`

inspec version: 3.9.3

Any idea ?

**Comment by aaronlippold** [comment]
Odd. What is the value in the inspec.yml file. Maybe I incorrectly set the
default value to a non-string and it's not even really getting to the
input.yaml file your passing.

**Comment by aaronlippold** [comment]
We may not be able to have nil string inputs in the inspec.yml file for
default values. We may have to standardize on default or XXX or something.

but the real expected behavior is that it throws an error saying the input
x is a required input and has no value, please provide a value for input x
before running the profile. So we should probably also open an issue as
well.

**Comment by aaronlippold** [comment]
Actually, I do see the issue, in the inspec.yml file we are pulling a lot of the values from ENV VARS, you need to make sure to set them before running the profile.

You will see: `<%=ENV['SOME_THING'] %>` which needs to be set in the environment.

**Comment by aaronlippold** [comment]
Reference L120 in the README.md

**Comment by sharmay** [comment]
> Actually, I do see the issue, in the inspec.yml file we are pulling a lot of the values from ENV VARS, you need to make sure to set them before running the profile.
> 
> You will see: `<%=ENV['SOME_THING'] %>` which needs to be set in the environment.

Why to use ENV vars and not values from input yaml ?

**Comment by aaronlippold** [comment]
More secure for sensitive information.

**Comment by hunleyd** [comment]
It looks like having nil env vars is  a no-go in my testing. And I'm also getting `Type 'Number' is not a valid input type.`  but i'm not sure what it thinks is a number? I created stig.env as:
```
export PG_OWNER=postgres
export PG_OWNER_GRP=postgres
export PG_OWNER_PWD=secret
export PG_DBA=postgres
export PG_DBA_PWD=secret
export PG_HOST="127.0.0.1"
export PG_USER=postgres
export PG_USER_PWD=secret
export PG_PORT=5432
export LOGIN_USER=postgres
export LOGIN_HOST="127.0.0.1"
export PG_SYSLOG_OWNER=postgres
```

which I then `source` before running `inspec exec pgstigcheck-inspec --attrs pgstigcheck-inspec/attributes.yml --reporter=json > inspec-report.json` ..  As you can see, I tried quoting the IP address thinking that was the culprit, but it's not (I originally had both LOGIN vars empty since I'm running this locally, that's when I got the 'nil' error).

**Comment by aaronlippold** [comment]
Covered in #28

**Pull request closed without merging**